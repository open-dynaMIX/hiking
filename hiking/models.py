import datetime
import os
from typing import List

import gpxpy
from sqlalchemy import (
    Column,
    Date,
    Float,
    Integer,
    Interval,
    String,
    Text,
    create_engine,
)
from sqlalchemy.orm import declarative_base, sessionmaker

from hiking.utils import DB_PATH, format_value, pretty_timedelta

Base = declarative_base()
engine = create_engine(
    f"sqlite:///{str(DB_PATH.absolute())}"
    if not os.environ.get("HIKING_TEST")
    else "sqlite://"  # , echo=True
)
Session = sessionmaker(bind=engine)
session = Session()


def create_tables():
    Base.metadata.create_all(engine)


class Hike(Base):
    __tablename__ = "hikes"

    id = Column(Integer, primary_key=True)
    name = Column(String, nullable=False)
    body = Column(Text, nullable=True)
    date = Column(Date, nullable=False)
    distance = Column(Float, nullable=False)
    elevation_gain = Column(Integer, nullable=False)
    elevation_loss = Column(Integer, nullable=False)
    duration = Column(Interval, nullable=False)
    gpx_xml = Column(Text)

    FIELD_PROPS = {
        "id": {
            "pretty_name": "ID",
            "supported_calculations": [],
            "data_view": True,
            "calculated_value": False,
        },
        "date": {
            "pretty_name": "Date",
            "supported_calculations": [],
            "data_view": True,
            "calculated_value": False,
        },
        "name": {
            "pretty_name": "Name",
            "supported_calculations": [],
            "data_view": True,
            "calculated_value": False,
        },
        "body": {
            "pretty_name": "Body",
            "supported_calculations": [],
            "data_view": False,
            "calculated_value": False,
        },
        "distance": {
            "pretty_name": "➡ km",
            "supported_calculations": ["sum", "avg", "min", "max"],
            "data_view": True,
            "calculated_value": False,
        },
        "elevation_gain": {
            "pretty_name": "⬈ m",
            "supported_calculations": ["sum", "avg", "min", "max"],
            "data_view": True,
            "calculated_value": False,
        },
        "elevation_loss": {
            "pretty_name": "⬊ m",
            "supported_calculations": ["sum", "avg", "min", "max"],
            "data_view": True,
            "calculated_value": False,
        },
        "duration": {
            "pretty_name": "⏱ ",
            "supported_calculations": ["sum", "avg", "min", "max"],
            "data_view": True,
            "calculated_value": False,
        },
        "speed": {
            "pretty_name": "km/h",
            "supported_calculations": ["avg", "min", "max"],
            "data_view": True,
            "calculated_value": True,
        },
        "gpx": {
            "pretty_name": "GPX",
            "supported_calculations": [],
            "data_view": False,
            "calculated_value": False,
        },
    }

    _gpx = None

    @property
    def speed(self):
        return self.distance * 60 / (self.duration.seconds / 60)

    @property
    def gpx(self):
        if self.gpx_xml and not self._gpx:
            self._gpx = gpxpy.parse(self.gpx_xml)
        return self._gpx

    def get_pretty_value(
        self,
        attr: str,
    ):
        value = getattr(self, attr)
        return format_value(value, attr)

    def get_stats(self) -> List[str]:
        serialized = []
        for attr, config in self.FIELD_PROPS.items():
            if not config["data_view"]:
                continue

            value = self.get_pretty_value(attr)
            serialized.append(value)

        return serialized

    def get_detail_stats(self) -> dict:
        serialized = {}
        for f, config in self.FIELD_PROPS.items():
            if not config["data_view"]:
                continue
            value = getattr(self, f)
            if isinstance(value, datetime.timedelta):
                value = pretty_timedelta(value)
            elif f == "speed":
                value = round(value, 2)
            serialized[config["pretty_name"]] = value

        return serialized

    def load_gpx(self, gpx: str):
        gpx_obj = gpxpy.parse(gpx)

        data = {
            "date": gpx_obj.time.date() if gpx_obj.time else None,
            "name": gpx_obj.name,
            "distance": round(gpx_obj.length_3d() / 1000, 2),
            "elevation_gain": round(gpx_obj.get_uphill_downhill().uphill),
            "elevation_loss": round(gpx_obj.get_uphill_downhill().downhill),
            "duration": gpx_obj.get_duration(),
            "gpx_xml": gpx,
        }
        for attr, value in data.items():
            if value:
                setattr(self, attr, value)

    def save(self):
        if self.id is None:
            session.add(self)
        return session.commit()

    def delete(self):
        session.delete(self)
        return session.commit()

    def __str__(self) -> str:
        return f"<Hike - {self.name} - {self.date} - {self.distance}km>"

    def __repr__(self) -> str:
        return self.__str__()
