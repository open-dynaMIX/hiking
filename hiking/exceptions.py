class HikingJsonLoaderException(Exception):
    pass


class HikingException(Exception):
    pass
