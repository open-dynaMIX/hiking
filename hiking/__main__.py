#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import logging
import sys

from hiking import commands
from hiking.arg_parsing import parse_arguments
from hiking.exceptions import HikingException, HikingJsonLoaderException
from hiking.import_export import JSON_IMPORT_EXAMPLE
from hiking.models import create_tables
from hiking.utils import DATA_HOME

logger = logging.getLogger(__name__)


def main():
    args = parse_arguments(sys.argv[1:])

    DATA_HOME.mkdir(exist_ok=True)
    create_tables()

    try:
        match args.command:
            case "show":
                commands.command_show(
                    args.ids,
                    args.daterange,
                    args.table_style,
                    args.order_key,
                    args.plot,
                )
            case "create" | "edit":
                commands.command_create_edit(pk=getattr(args, "id", None), gpx=args.gpx)
            case "delete":
                commands.command_delete(args.ids, args.all, args.force, args.quiet)
            case "import":
                commands.command_import(args.json_data)
            case "export":
                commands.command_export(
                    args.export_dir, args.ids, args.daterange, args.include_ids
                )

    except HikingJsonLoaderException as e:
        logger.warning(
            f"Invalid data in hiking.json: {e.args[0]}\n\nExpected format:\n{JSON_IMPORT_EXAMPLE}"
        )
    except HikingException as e:
        logger.error(f"Error: {e}")
    except (KeyboardInterrupt, EOFError):  # pragma: no cover
        sys.exit(0)


if __name__ == "__main__":  # pragma: no cover
    main()
