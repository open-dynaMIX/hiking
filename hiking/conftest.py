from pathlib import Path

import pytest
from pytest_factoryboy.fixture import register

from hiking import factories
from hiking.models import create_tables, engine, session

OWN_DIR = Path(__file__).resolve().parent
GPX_TEST_FILE = OWN_DIR / "tests" / "data" / "test.gpx"


register(factories.HikeFactory)


@pytest.fixture(scope="session", autouse=True)
def setup_db():
    create_tables()


@pytest.fixture(autouse=True)
def db():
    yield
    with engine.connect() as con:
        con.execute("DELETE FROM hikes;")
    session.commit()
    session.close()
    session.begin()


@pytest.fixture
def gpx_xml():
    with GPX_TEST_FILE.open("r") as f:
        xml = f.read()
    return xml
