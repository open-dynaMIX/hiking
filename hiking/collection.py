import datetime
from dataclasses import dataclass
from typing import List, Tuple, Union

from sqlalchemy import Interval, func
from sqlalchemy.orm import Query

from hiking.db_utils import session
from hiking.models import Hike, get_filtered_query
from hiking.utils import SlimDateRange, format_value


@dataclass
class HikeCollection:
    hikes: Query

    def get_hikes_attr_list(
        self, attr: str
    ) -> List[Union[float, int, datetime.date, datetime.timedelta]]:
        return [getattr(hike, attr) for hike in self.hikes.all()]

    def sum(
        self, attr: str = "distance"
    ) -> Union[float, int, datetime.date, datetime.timedelta]:
        is_interval = isinstance(getattr(Hike, attr).expression.type, Interval)
        if Hike.FIELD_PROPS[attr]["calculated_value"] or is_interval:
            attr_list = self.get_hikes_attr_list(attr)
            if is_interval:
                return sum(attr_list, datetime.timedelta())

            return sum(attr_list)  # pragma: no cover

        result = session.query(func.sum(getattr(Hike, attr))).first()[0]
        return result

    def avg(self, attr: str = "distance") -> Union[float, datetime.timedelta]:
        if attr == "duration":
            return (
                sum([h.duration for h in self.hikes.all()], datetime.timedelta())
                / self.hikes.count()
            )
        return sum(getattr(h, attr) for h in self.hikes.all()) / self.hikes.count()

    def max(
        self, attr: str = "distance"
    ) -> Union[float, int, datetime.date, datetime.timedelta]:
        if attr == "speed":
            return sorted(self.hikes.all(), key=lambda x: x.speed, reverse=True)[
                0
            ].speed

        result = getattr(
            self.hikes.order_by(getattr(Hike, attr).desc()).limit(1).first(), attr
        )
        return result

    def min(
        self, attr: str = "distance"
    ) -> Union[float, int, datetime.date, datetime.timedelta]:
        if attr == "speed":
            return sorted(self.hikes.all(), key=lambda x: x.speed)[0].speed

        result = getattr(
            self.hikes.order_by(getattr(Hike, attr).asc()).limit(1).first(), attr
        )
        return result

    def get_hikes_stats(self, order_params: Tuple[str, bool]) -> List[List[str]]:
        if order_params[0] == "speed":
            hike_list = sorted(
                self.hikes.all(), key=lambda x: x.speed, reverse=order_params[1]
            )
            return [hike.get_stats() for hike in hike_list]

        order = getattr(Hike, order_params[0])
        if order_params[1]:
            order = order.desc()
        query = self.hikes.order_by(order)
        return [hike.get_stats() for hike in query]

    def calc_and_format_value(self, calc: str, attr: str) -> str:
        result = getattr(self, calc)(attr)
        return format_value(result, attr)

    def get_totals(self) -> List[str]:
        def get_summary_cell(attr: str, supported_calculations: List[str]):
            if supported_calculations:
                cell = {}
                for calc, pretty_calc in [
                    ("sum", "Σ "),
                    ("avg", "⌀ "),
                    ("max", "↑ "),
                    ("min", "↓ "),
                ]:
                    cell[pretty_calc] = "-"
                    if calc in supported_calculations:
                        cell[pretty_calc] = self.calc_and_format_value(calc, attr)
                return cell

        d = [
            "",
            "STATS",
            str(self.hikes.count()),
            *[
                get_summary_cell(attr, config["supported_calculations"])
                for attr, config in Hike.FIELD_PROPS.items()
                if config["supported_calculations"]
            ],
        ]
        return d

    def get_collection_stats(
        self, order_params: Tuple[str, bool], add_totals: bool = True
    ) -> Tuple[List, List]:
        stats = self.get_hikes_stats(order_params)

        result = [
            *stats,
        ]

        footer = None
        if add_totals and len(stats) > 1:
            # Only add totals if more than one hike is present
            footer = self.get_totals()

        return result, footer

    def __str__(self) -> str:
        return (
            f"<HikeCollection - "
            f"containing {self.hikes.count()} hikes - "
            f"total {self.sum()}km>"
        )

    def __repr__(self) -> str:
        return self.__str__()


def get_collection(
    ids: List[int],
    daterange: "SlimDateRange",
):
    query = get_filtered_query(ids, daterange)
    collection = HikeCollection(hikes=query)
    return collection
