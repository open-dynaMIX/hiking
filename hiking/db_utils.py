from typing import List

from hiking.collection import HikeCollection
from hiking.models import Hike, session
from hiking.utils import SlimDateRange


def get_filtered_query(
    ids: List[int],
    daterange: "SlimDateRange",
):
    query = (
        session.query(Hike)
        .filter(Hike.date >= daterange.lower)
        .filter(Hike.date <= daterange.upper)
    )
    if ids:
        query = query.filter(Hike.id.in_(ids))
    return query


def get_collection(
    ids: List[int],
    daterange: "SlimDateRange",
):
    query = get_filtered_query(ids, daterange)
    collection = HikeCollection(hikes=query, session=session)
    return collection
