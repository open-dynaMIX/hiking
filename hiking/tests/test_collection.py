import pytest

from hiking.models import Hike


@pytest.mark.parametrize("field", [f.info for f in Hike.FIELDS])
def test_collection_attr_functions(collection, snapshot, field):
    assert collection.get_hikes_attr_list(field["name"]) == snapshot(
        name=f"{field['name']} - attr_list"
    )

    for calc in ["sum", "avg", "min", "max"]:
        if calc in field["supported_calculations"]:
            assert getattr(collection, calc)(field["name"]) == snapshot(
                name=f"{field['name']} - {calc} - raw"
            )
            assert collection.calc_and_format_value(calc, field["name"]) == snapshot(
                name=f"{field['name']} - {calc} - pretty"
            )


@pytest.mark.parametrize(
    "order_params",
    [
        ("date", False),
        ("date", True),
        ("elevation_gain", False),
        ("elevation_gain", True),
    ],
)
def test_collection_hikes_stats(
    collection, snapshot, caplog, debug_logging, order_params
):
    stats = collection.get_hikes_stats(order_params)
    # remove the IDs for comparison
    for hike in stats:
        del hike[0]

    prefix = "-" if order_params[1] else ""
    assert stats == snapshot(name=f'order_param: "{prefix}{order_params[0]} - stats"')

    assert len(caplog.messages) == 3
    assert caplog.messages[0] == "BEGIN (implicit)"
    assert caplog.messages[2].startswith("[generated in 0.00")
    assert caplog.messages[1] == snapshot(
        name=f'order_param: "{prefix}{order_params[0]}" - query'
    )


def test_collection_get_totals(collection, snapshot):
    assert collection.get_totals() == snapshot


@pytest.mark.parametrize(
    "order_params",
    [
        ("date", False),
        ("date", True),
        ("elevation_gain", False),
        ("elevation_gain", True),
    ],
)
def test_collection_collection_stats(collection, snapshot, order_params):
    stats, footer = collection.get_collection_stats(order_params)
    # remove the IDs for comparison
    for hike in stats:
        del hike[0]

    prefix = "-" if order_params[1] else ""

    assert stats == snapshot(name=f'order_param: "{prefix}{order_params[0]}" - stats')
    assert footer == snapshot(name=f'order_param: "{prefix}{order_params[0]}" - footer')
