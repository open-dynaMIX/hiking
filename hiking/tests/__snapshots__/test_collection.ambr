# serializer version: 1
# name: test_collection_attr_functions[field0][id - attr_list]
  list([
    1,
    2,
    3,
  ])
# ---
# name: test_collection_attr_functions[field1][date - attr_list]
  list([
    datetime.date(2020, 5, 16),
    datetime.date(2005, 4, 5),
    datetime.date(2024, 2, 27),
  ])
# ---
# name: test_collection_attr_functions[field2][name - attr_list]
  list([
    'Paul Diaz',
    'Gina Lynch',
    'John Rhodes',
  ])
# ---
# name: test_collection_attr_functions[field3][body - attr_list]
  list([
    'Owner candidate condition which central. Owner support record upon election. Pattern pick camera benefit. Poor health purpose practice.',
    'Summer best manager up. Without hit science onto practice federal. Fight type street.',
    'International music call field door prove. Hit treat garden become a art. Like learn short present store choose deep.',
  ])
# ---
# name: test_collection_attr_functions[field4][distance - attr_list]
  list([
    17.43,
    17.8,
    14.95,
  ])
# ---
# name: test_collection_attr_functions[field4][distance - avg - pretty]
  '16.7'
# ---
# name: test_collection_attr_functions[field4][distance - avg - raw]
  16.726666666666667
# ---
# name: test_collection_attr_functions[field4][distance - max - pretty]
  '17.8'
# ---
# name: test_collection_attr_functions[field4][distance - max - raw]
  17.8
# ---
# name: test_collection_attr_functions[field4][distance - min - pretty]
  '14.9'
# ---
# name: test_collection_attr_functions[field4][distance - min - raw]
  14.95
# ---
# name: test_collection_attr_functions[field4][distance - sum - pretty]
  '50.2'
# ---
# name: test_collection_attr_functions[field4][distance - sum - raw]
  50.18
# ---
# name: test_collection_attr_functions[field5][elevation_gain - attr_list]
  list([
    3268,
    2732,
    2452,
  ])
# ---
# name: test_collection_attr_functions[field5][elevation_gain - avg - pretty]
  '2817'
# ---
# name: test_collection_attr_functions[field5][elevation_gain - avg - raw]
  2817.3333333333335
# ---
# name: test_collection_attr_functions[field5][elevation_gain - max - pretty]
  '3268'
# ---
# name: test_collection_attr_functions[field5][elevation_gain - max - raw]
  3268
# ---
# name: test_collection_attr_functions[field5][elevation_gain - min - pretty]
  '2452'
# ---
# name: test_collection_attr_functions[field5][elevation_gain - min - raw]
  2452
# ---
# name: test_collection_attr_functions[field5][elevation_gain - sum - pretty]
  '8452'
# ---
# name: test_collection_attr_functions[field5][elevation_gain - sum - raw]
  8452
# ---
# name: test_collection_attr_functions[field6][elevation_loss - attr_list]
  list([
    2652,
    3346,
    733,
  ])
# ---
# name: test_collection_attr_functions[field6][elevation_loss - avg - pretty]
  '2244'
# ---
# name: test_collection_attr_functions[field6][elevation_loss - avg - raw]
  2243.6666666666665
# ---
# name: test_collection_attr_functions[field6][elevation_loss - max - pretty]
  '3346'
# ---
# name: test_collection_attr_functions[field6][elevation_loss - max - raw]
  3346
# ---
# name: test_collection_attr_functions[field6][elevation_loss - min - pretty]
  '733'
# ---
# name: test_collection_attr_functions[field6][elevation_loss - min - raw]
  733
# ---
# name: test_collection_attr_functions[field6][elevation_loss - sum - pretty]
  '6731'
# ---
# name: test_collection_attr_functions[field6][elevation_loss - sum - raw]
  6731
# ---
# name: test_collection_attr_functions[field7][duration - attr_list]
  list([
    datetime.timedelta(seconds=13843, microseconds=990515),
    datetime.timedelta(seconds=12506, microseconds=309075),
    datetime.timedelta(seconds=11832, microseconds=211302),
  ])
# ---
# name: test_collection_attr_functions[field7][duration - avg - pretty]
  '03:32'
# ---
# name: test_collection_attr_functions[field7][duration - avg - raw]
  datetime.timedelta(seconds=12727, microseconds=503631)
# ---
# name: test_collection_attr_functions[field7][duration - max - pretty]
  '03:50'
# ---
# name: test_collection_attr_functions[field7][duration - max - raw]
  datetime.timedelta(seconds=13843, microseconds=990515)
# ---
# name: test_collection_attr_functions[field7][duration - min - pretty]
  '03:17'
# ---
# name: test_collection_attr_functions[field7][duration - min - raw]
  datetime.timedelta(seconds=11832, microseconds=211302)
# ---
# name: test_collection_attr_functions[field7][duration - sum - pretty]
  '10:36'
# ---
# name: test_collection_attr_functions[field7][duration - sum - raw]
  datetime.timedelta(seconds=38182, microseconds=510892)
# ---
# name: test_collection_attr_functions[field8][gpx - attr_list]
  list([
    None,
    None,
    None,
  ])
# ---
# name: test_collection_attr_functions[field9][speed - attr_list]
  list([
    4.532832478508993,
    5.123940508555893,
    4.54868154158215,
  ])
# ---
# name: test_collection_attr_functions[field9][speed - avg - pretty]
  '4.74'
# ---
# name: test_collection_attr_functions[field9][speed - avg - raw]
  4.735151509549012
# ---
# name: test_collection_attr_functions[field9][speed - max - pretty]
  '5.12'
# ---
# name: test_collection_attr_functions[field9][speed - max - raw]
  5.123940508555893
# ---
# name: test_collection_attr_functions[field9][speed - min - pretty]
  '4.53'
# ---
# name: test_collection_attr_functions[field9][speed - min - raw]
  4.532832478508993
# ---
# name: test_collection_collection_stats[order_params0][order_param: "date" - footer]
  list([
    '',
    'STATS',
    '3',
    dict({
      'Σ ': '50.2',
      '↑ ': '17.8',
      '↓ ': '14.9',
      '⌀ ': '16.7',
    }),
    dict({
      'Σ ': '8452',
      '↑ ': '3268',
      '↓ ': '2452',
      '⌀ ': '2817',
    }),
    dict({
      'Σ ': '6731',
      '↑ ': '3346',
      '↓ ': '733',
      '⌀ ': '2244',
    }),
    dict({
      'Σ ': '10:36',
      '↑ ': '03:50',
      '↓ ': '03:17',
      '⌀ ': '03:32',
    }),
    dict({
      'Σ ': '-',
      '↑ ': '5.12',
      '↓ ': '4.53',
      '⌀ ': '4.74',
    }),
  ])
# ---
# name: test_collection_collection_stats[order_params0][order_param: "date" - stats]
  list([
    list([
      '2',
      '2005-04-05',
      'Gina Lynch',
      '17.8',
      '2732',
      '3346',
      '03:28',
      '5.12',
    ]),
    list([
      '1',
      '2020-05-16',
      'Paul Diaz',
      '17.4',
      '3268',
      '2652',
      '03:50',
      '4.53',
    ]),
    list([
      '3',
      '2024-02-27',
      'John Rhodes',
      '14.9',
      '2452',
      '733',
      '03:17',
      '4.55',
    ]),
  ])
# ---
# name: test_collection_collection_stats[order_params1][order_param: "-date" - footer]
  list([
    '',
    'STATS',
    '3',
    dict({
      'Σ ': '50.2',
      '↑ ': '17.8',
      '↓ ': '14.9',
      '⌀ ': '16.7',
    }),
    dict({
      'Σ ': '8452',
      '↑ ': '3268',
      '↓ ': '2452',
      '⌀ ': '2817',
    }),
    dict({
      'Σ ': '6731',
      '↑ ': '3346',
      '↓ ': '733',
      '⌀ ': '2244',
    }),
    dict({
      'Σ ': '10:36',
      '↑ ': '03:50',
      '↓ ': '03:17',
      '⌀ ': '03:32',
    }),
    dict({
      'Σ ': '-',
      '↑ ': '5.12',
      '↓ ': '4.53',
      '⌀ ': '4.74',
    }),
  ])
# ---
# name: test_collection_collection_stats[order_params1][order_param: "-date" - stats]
  list([
    list([
      '3',
      '2024-02-27',
      'John Rhodes',
      '14.9',
      '2452',
      '733',
      '03:17',
      '4.55',
    ]),
    list([
      '1',
      '2020-05-16',
      'Paul Diaz',
      '17.4',
      '3268',
      '2652',
      '03:50',
      '4.53',
    ]),
    list([
      '2',
      '2005-04-05',
      'Gina Lynch',
      '17.8',
      '2732',
      '3346',
      '03:28',
      '5.12',
    ]),
  ])
# ---
# name: test_collection_collection_stats[order_params2][order_param: "elevation_gain" - footer]
  list([
    '',
    'STATS',
    '3',
    dict({
      'Σ ': '50.2',
      '↑ ': '17.8',
      '↓ ': '14.9',
      '⌀ ': '16.7',
    }),
    dict({
      'Σ ': '8452',
      '↑ ': '3268',
      '↓ ': '2452',
      '⌀ ': '2817',
    }),
    dict({
      'Σ ': '6731',
      '↑ ': '3346',
      '↓ ': '733',
      '⌀ ': '2244',
    }),
    dict({
      'Σ ': '10:36',
      '↑ ': '03:50',
      '↓ ': '03:17',
      '⌀ ': '03:32',
    }),
    dict({
      'Σ ': '-',
      '↑ ': '5.12',
      '↓ ': '4.53',
      '⌀ ': '4.74',
    }),
  ])
# ---
# name: test_collection_collection_stats[order_params2][order_param: "elevation_gain" - stats]
  list([
    list([
      '3',
      '2024-02-27',
      'John Rhodes',
      '14.9',
      '2452',
      '733',
      '03:17',
      '4.55',
    ]),
    list([
      '2',
      '2005-04-05',
      'Gina Lynch',
      '17.8',
      '2732',
      '3346',
      '03:28',
      '5.12',
    ]),
    list([
      '1',
      '2020-05-16',
      'Paul Diaz',
      '17.4',
      '3268',
      '2652',
      '03:50',
      '4.53',
    ]),
  ])
# ---
# name: test_collection_collection_stats[order_params3][order_param: "-elevation_gain" - footer]
  list([
    '',
    'STATS',
    '3',
    dict({
      'Σ ': '50.2',
      '↑ ': '17.8',
      '↓ ': '14.9',
      '⌀ ': '16.7',
    }),
    dict({
      'Σ ': '8452',
      '↑ ': '3268',
      '↓ ': '2452',
      '⌀ ': '2817',
    }),
    dict({
      'Σ ': '6731',
      '↑ ': '3346',
      '↓ ': '733',
      '⌀ ': '2244',
    }),
    dict({
      'Σ ': '10:36',
      '↑ ': '03:50',
      '↓ ': '03:17',
      '⌀ ': '03:32',
    }),
    dict({
      'Σ ': '-',
      '↑ ': '5.12',
      '↓ ': '4.53',
      '⌀ ': '4.74',
    }),
  ])
# ---
# name: test_collection_collection_stats[order_params3][order_param: "-elevation_gain" - stats]
  list([
    list([
      '1',
      '2020-05-16',
      'Paul Diaz',
      '17.4',
      '3268',
      '2652',
      '03:50',
      '4.53',
    ]),
    list([
      '2',
      '2005-04-05',
      'Gina Lynch',
      '17.8',
      '2732',
      '3346',
      '03:28',
      '5.12',
    ]),
    list([
      '3',
      '2024-02-27',
      'John Rhodes',
      '14.9',
      '2452',
      '733',
      '03:17',
      '4.55',
    ]),
  ])
# ---
# name: test_collection_get_totals
  list([
    '',
    'STATS',
    '3',
    dict({
      'Σ ': '50.2',
      '↑ ': '17.8',
      '↓ ': '14.9',
      '⌀ ': '16.7',
    }),
    dict({
      'Σ ': '8452',
      '↑ ': '3268',
      '↓ ': '2452',
      '⌀ ': '2817',
    }),
    dict({
      'Σ ': '6731',
      '↑ ': '3346',
      '↓ ': '733',
      '⌀ ': '2244',
    }),
    dict({
      'Σ ': '10:36',
      '↑ ': '03:50',
      '↓ ': '03:17',
      '⌀ ': '03:32',
    }),
    dict({
      'Σ ': '-',
      '↑ ': '5.12',
      '↓ ': '4.53',
      '⌀ ': '4.74',
    }),
  ])
# ---
# name: test_collection_hikes_stats[order_params0][order_param: "date - stats"]
  list([
    list([
      '2',
      '2005-04-05',
      'Gina Lynch',
      '17.8',
      '2732',
      '3346',
      '03:28',
      '5.12',
    ]),
    list([
      '1',
      '2020-05-16',
      'Paul Diaz',
      '17.4',
      '3268',
      '2652',
      '03:50',
      '4.53',
    ]),
    list([
      '3',
      '2024-02-27',
      'John Rhodes',
      '14.9',
      '2452',
      '733',
      '03:17',
      '4.55',
    ]),
  ])
# ---
# name: test_collection_hikes_stats[order_params0][order_param: "date" - query]
  '''
  SELECT hikes.id AS hikes_id, hikes.date AS hikes_date, hikes.name AS hikes_name, hikes.distance AS hikes_distance, hikes.elevation_gain AS hikes_elevation_gain, hikes.elevation_loss AS hikes_elevation_loss, hikes.duration AS hikes_duration 
  FROM hikes ORDER BY hikes.date
  '''
# ---
# name: test_collection_hikes_stats[order_params1][order_param: "-date - stats"]
  list([
    list([
      '3',
      '2024-02-27',
      'John Rhodes',
      '14.9',
      '2452',
      '733',
      '03:17',
      '4.55',
    ]),
    list([
      '1',
      '2020-05-16',
      'Paul Diaz',
      '17.4',
      '3268',
      '2652',
      '03:50',
      '4.53',
    ]),
    list([
      '2',
      '2005-04-05',
      'Gina Lynch',
      '17.8',
      '2732',
      '3346',
      '03:28',
      '5.12',
    ]),
  ])
# ---
# name: test_collection_hikes_stats[order_params1][order_param: "-date" - query]
  '''
  SELECT hikes.id AS hikes_id, hikes.date AS hikes_date, hikes.name AS hikes_name, hikes.distance AS hikes_distance, hikes.elevation_gain AS hikes_elevation_gain, hikes.elevation_loss AS hikes_elevation_loss, hikes.duration AS hikes_duration 
  FROM hikes ORDER BY hikes.date DESC
  '''
# ---
# name: test_collection_hikes_stats[order_params2][order_param: "elevation_gain - stats"]
  list([
    list([
      '3',
      '2024-02-27',
      'John Rhodes',
      '14.9',
      '2452',
      '733',
      '03:17',
      '4.55',
    ]),
    list([
      '2',
      '2005-04-05',
      'Gina Lynch',
      '17.8',
      '2732',
      '3346',
      '03:28',
      '5.12',
    ]),
    list([
      '1',
      '2020-05-16',
      'Paul Diaz',
      '17.4',
      '3268',
      '2652',
      '03:50',
      '4.53',
    ]),
  ])
# ---
# name: test_collection_hikes_stats[order_params2][order_param: "elevation_gain" - query]
  '''
  SELECT hikes.id AS hikes_id, hikes.date AS hikes_date, hikes.name AS hikes_name, hikes.distance AS hikes_distance, hikes.elevation_gain AS hikes_elevation_gain, hikes.elevation_loss AS hikes_elevation_loss, hikes.duration AS hikes_duration 
  FROM hikes ORDER BY hikes.elevation_gain
  '''
# ---
# name: test_collection_hikes_stats[order_params3][order_param: "-elevation_gain - stats"]
  list([
    list([
      '1',
      '2020-05-16',
      'Paul Diaz',
      '17.4',
      '3268',
      '2652',
      '03:50',
      '4.53',
    ]),
    list([
      '2',
      '2005-04-05',
      'Gina Lynch',
      '17.8',
      '2732',
      '3346',
      '03:28',
      '5.12',
    ]),
    list([
      '3',
      '2024-02-27',
      'John Rhodes',
      '14.9',
      '2452',
      '733',
      '03:17',
      '4.55',
    ]),
  ])
# ---
# name: test_collection_hikes_stats[order_params3][order_param: "-elevation_gain" - query]
  '''
  SELECT hikes.id AS hikes_id, hikes.date AS hikes_date, hikes.name AS hikes_name, hikes.distance AS hikes_distance, hikes.elevation_gain AS hikes_elevation_gain, hikes.elevation_loss AS hikes_elevation_loss, hikes.duration AS hikes_duration 
  FROM hikes ORDER BY hikes.elevation_gain DESC
  '''
# ---
